### PR TITLE
The unwinder of goto-instrument is finally enabled 

### PR DESCRIPTION
**Pull request by theyoucheng**
Note that previously the piece of codes for```--unwind``` in```goto-instrument``` was in fact disabled. This patch on one hand triggers the unwinding operation in ```goto-instrument```, on the other hand it makes implementation of  the unwinder of ```goto-instrument``` more completed. 

After unwinding the loop body for the specified number of times, there are now two choices. By default, the unwinder will insert an assume statement; instead, the type2 unwinder will choose to continue the loop.

### CODE REVIEW

**Comment by kroening** [comment]
Fix the conflicts, please!

**Comment by theyoucheng** [comment]
I later sent another pull request that was merged with #194 

This one should be closed (or deleted if possible). Sorry for the confusion.

File: src/goto-instrument/goto_instrument_parse_options.cpp [hunk 1]
```diff
@@ -144,6 +147,18 @@ int goto_instrument_parse_optionst::doit()
     get_goto_program();
     instrument_goto_program();
 
+    if(cmdline.isset("unwind"))
+    {
+
+      int k = stoi(cmdline.get_value("unwind"));
+
+      goto_unwind(goto_functions, k);
+      //goto_unwind_type2(goto_functions, k);
```
Comment (peterschrammel) [inline src/goto-instrument/goto_instrument_parse_options.cpp]: @theyoucheng, what is goto_unwind_type2 for?

**Comment by theyoucheng** [comment]
Hi Peter,

Considering the semantics of "unwinding" operations, several options are there. 

For example, 

1) after unwinding the loop body a number of times, simply breaking the loop (under-approximation)
2) after unwinding the loop body a number of times, continuing the loop (this is exact)

When I implemented the unwinder for goto-instrument, I played with both options, and "type2" is for the 2nd semantic. 

However, before reaching a clear agreement on how to utilize the unwinder of goto-instrument, I finally decided to only keep a single semantic in the implementation (i.e. the 1st one), for
the upgraded patch (#194) sent to the mainline CBMC.    

Best,
Youcheng

**Comment by martin-cs** [comment]
On Thu, 2016-08-25 at 01:19 -0700, Youcheng Sun wrote:
> Hi Peter,
> 
> Considering the semantics of "unwinding" operations, several options are there. 
> 
> For example, 
> 
> 1) after unwinding the loop body a number of times, simply breaking the loop (under-approximation)
> 2) after unwinding the loop body a number of times, continuing the loop (this is exact)
> 
> When I implemented the unwinder for goto-instrument, I played with both options, and "type2" is for the 2nd semantic. 
> 
> However, before reaching a clear agreement on how to utilize the unwinder of goto-instrument, I finally decided to only keep a single semantic in the implementation (i.e. the 1st one), for
> the upgraded patch (#194) sent to the mainline CBMC.    

Ideally we should have an argument that picks which one of the four
semantics we want:

1. Unwind N times, then stop (this is what --partial-loops does)
2. Unwind N times, then include the loop
3. Unwind N times, then:
__CPROVER_assert(loop_condition, "Static unwinding assertion $LOOP_ID")
4. Unwind N times, then:
__CPROVER_assume(!loop_condition)

Cheers,
 - Martin

**Comment by peterschrammel** [comment]
Yes, @theyoucheng, we should add the selector of different semantics that @martin-cs proposes. 
For my purposes, I need 2.

**Comment by martin-cs** [comment]
On Sat, 2016-09-17 at 10:22 -0700, Peter Schrammel wrote:
> Yes, @theyoucheng, we should add the selector of different semantics that @martin-cs proposes. 
> For my purposes, I need 2.

Yeah; we need --unwind-set as well for a project.

Youcheng?  Would you mind doing these when you have a moment?

Cheers,
 - Martin

**Pull request closed without merging**